### PR TITLE
fix(material/core): add checkmark for single-select

### PR DESCRIPTION
**Pull request by zarend**
Add checkmark to mat-option for single-select. Fix a11y issues where selected state is visually communicated with color alone. Communicate selection with both color and a checkmark indicator. Affect components that use mat-option for single-selection, which include select and autocomplete.

Add an `appearance` Input to mat-pseudo-checkbox. "full" appearance renders a checkbox, which is the current behavior. Render "full" appearance by default. "minimal" apperance renders only a checkmark.

Summary of API and behavior changes:
 - Add an `@Input appearance` to pseudo-checkbox with options for "full" and "minimal".
 - mat-option renders "minimal" appearance for single-select, which looks like a checkmark.
 - Select and autocomplete components display checkmark on selected option.

What's not changing:
 - does not affect multiple selection
 - mat-option renders "full" apperance by default, which is same as existing behavior

Fixes: #25961

image
image
image

### CODE REVIEW

File: src/material/core/option/option.html [hunk 1]
```diff
@@ -1,5 +1,6 @@
-<mat-pseudo-checkbox *ngIf="multiple" class="mat-mdc-option-pseudo-checkbox"
```
[the remark this answers is not in the document]
Comment (zarend) [reply src/material/core/option/option.html:1]: Putting the checkmark on the end would look reasonable to me. I think it would make more sense if there was a API to configure if it's on the left or the right, like how we do it for MDC List. That's because it seems to me like it depends on the angular app's design and personal preference. WDYT?

[the remark this answers is not in the document]
Comment (crisbeto) [reply src/material/core/option/option.html:1]: I suspect that it defaults to `accent`, because `mat-checkbox` does too and it was written to match `mat-checkbox` as closely as possible. We could leave on the default and have the consuming component override the color. 

That being said, I found the current color a little weird since your select screenshot looks right, but the autocomplete does not.

[the remark this answers is not in the document]
Comment (zarend) [reply src/material/core/option/option.html:1]: I changed this to use a mathematic formula to place the checkmark. Hopefully this fixes the centering. Does it look centered to you? I can barely tell the difference between before and after.

[the remark this answers is not in the document]
Comment (zarend) [reply src/material/core/option/option.html:1]: For 2, I sent out #25983 to fix autocomplete not applying the right theme to the panel.

[the remark this answers is not in the document]
Comment (zarend) [reply src/material/core/option/option.html:1]: Okay, I would consider that out-of-scope of this PR, but I filed a feature request for it, so we can track it. #25984

**Comment by zarend** [comment]
rebased with upstream/main. Maybe this will fix the issue with dev-app deployment 🤞

File: src/material/core/option/option.html (hunk 1 above)
[the remark this answers is not in the document]
Comment (mmalerba) [reply src/material/core/option/option.html:1]: I like having the check at the start by default, matches the native select behavior

[the remark this answers is not in the document]
Comment (mmalerba) [reply src/material/core/option/option.html:1]: does this PR include the changes from the PR you linked? I still see the autocomplete check showing up as the wrong color

File: src/material/core/selection/pseudo-checkbox/_pseudo-checkbox-common.scss [hunk 2]
```diff
@@ -4,26 +4,79 @@
...
+  width: $width;
+  height: $height;
+
+  // Rotate on the center of the element. This makes it easier to center the checkmark graphic.
+  transform-origin: center;
+
+  // Take negative one times the distance from the top corner of the checkmark graphic to the top
+  // of the element in its rotated position. This accounts for the top corner of the elemant being
+  // blank since we only use the left and bottom borders to draw the checkmark graphic.
+  top: -1 * math.div($short-side - checkbox-common.$border-width, math.sqrt(2));
```
Comment (mmalerba) [inline src/material/core/selection/pseudo-checkbox/_pseudo-checkbox-common.scss:29]: maths 😎

**Review by mmalerba — COMMENTED** [review]
LGTM aside from the color issue

File: src/material/core/option/option.html (hunk 1 above)
[the remark this answers is not in the document]
Comment (crisbeto) [reply src/material/core/option/option.html:1]: To clarify what I meant above: I think that we should keep the checkmark at the end initially, but I'm fine with introducing an API to change it later on. I think that having it in the beginning looks weird, because it also introduces a whitespace before deselected options which almost makes it look like we were trying to center the text.

image

**Review by crisbeto — CHANGES_REQUESTED** [review]
https://github.com/angular/components/pull/25962#discussion_r1026167832 should be resolved before we move forward.

File: src/material/core/option/option.html (hunk 1 above)
[the remark this answers is not in the document]
Comment (mmalerba) [reply src/material/core/option/option.html:1]: Ah yeah, I guess native select avoids this situation since there's always a selected element

**Comment by zarend** [comment]
image
The autocomplete theming fix has landed. I rebased this branch, and we can see that the colors are corrected.

File: src/dev-app/select/select-demo.html [hunk 3]
```diff
@@ -12,8 +12,8 @@
         <mat-label>Drink</mat-label>
         <mat-select [(ngModel)]="currentDrink" [required]="drinksRequired"
           [disabled]="drinksDisabled" #drinkControl="ngModel">
-          <mat-option>None</mat-option>
-          <mat-option *ngFor="let drink of drinks" [value]="drink.value" [disabled]="drink.disabled">
+          <mat-option value="" [disabled]="drinksOptionsDisabled">None</mat-option>
```
Comment (crisbeto) [inline src/dev-app/select/select-demo.html:15]: Can you revert this change to `value=""`. This option is here so that it's easier to test options without a `value`.

**Review by crisbeto — APPROVED** [review]
One final nit, otherwise LGTM

File: src/material/core/option/option.html (hunk 1 above)
[the remark this answers is not in the document]
Comment (zarend) [reply src/material/core/option/option.html]: Done, minimal checkbox is at the end of the option.

[the remark this answers is not in the document]
Comment (zarend) [reply src/material/core/option/option.html]: #25983 landed, and I believe the appearance for the autocomplete has been corrected. Could someone please confirm that this is the correct appearance? Is the font color supposed to change when there's a checkmark? See attached screenshots of both with and without selection indicators.

image
image

**Comment by zarend** [comment]
@mmalerba @crisbeto could you review this again please as it has changed. Since last reviewed, added opt-outs for Select and Autocomplete.

I have a few questions:
 - Do we need an opt-out for <mat-option/> when used outside of select or autocomplete? As implemented, the opt-out only affect mat-option inside of a Select or Autocomplete. For ripple, there's a way to disable it from the Select/Autocomplete and from the mat-option too. Should hideSingleSelectionIndicator do the same?
 - Could you take a look at `_syncParentProperties`? I ran into an issue where autocomplete wouldn't work when testing on the autocomplete overview demo (`/autocomplete`). Clicking the "Hide Single-Selection Indicators" didn't seem to do anything. It seemed like change detection wasn't running because we mutated the `_parent` property on MatOption and it uses on-push change detection. I couldn't repro on Select, but I did the same fix for Select anyways.
  - For `MatAutocomplete`, would it be better to this another way, like assigning `defaults` to the base class, or using `inject` function? That way we don't have the change the constructor of the derived class?

File: src/material/autocomplete/autocomplete.ts [hunk 4]
```diff
@@ -336,4 +343,37 @@ export class MatAutocomplete extends _MatAutocompleteBase {
...
+  }
+
+  /** Whether checkmark indicator for single-selection options is hidden. */
+  @Input()
+  get hideSingleSelectionIndicator(): boolean {
+    return this._hideSingleSelectionIndicator;
+  }
+  set hideSingleSelectionIndicator(value: BooleanInput) {
+    this._hideSingleSelectionIndicator = coerceBooleanProperty(value);
+    this._syncParentProperties();
```
Comment (crisbeto) [inline src/material/autocomplete/autocomplete.ts:365]: In what case does this become a problem? E.g. is it only if the binding changes while the autocomplete is open or does it not get picked up on open either?

Comment (zarend) [reply src/material/autocomplete/autocomplete.ts:365]: It doesn't get picked up on open. Although, it does get picked up when selecting an option, selecting the same option again, or selecting a different option. It seems to me that selecting an option triggers chance detection by conincidence, which cases the child component to detect the change to `hideSingleSelectionIndicator`.

```
// option.ts
  /** Whether to display checkmark for single-selection. */
  get hideSingleSelectionIndicator(): boolean {
    return !!(this._parent && this._parent.hideSingleSelectionIndicator);
  }
```

Both MatOption and MatAutocomplete use on-push change detection. I believe that means that mutating `_parent` will not trigger change detection on MatOption. That's why we have to manually call `markForCheck`. Am I understanding the change detection correctly :)?

Here's a video capture of the behavior I see when I remove `_syncParentProperties`.


https://user-images.githubusercontent.com/7720245/210853511-6c112c5b-36f8-4e15-ac82-ffe0307f65c1.mov

Comment (crisbeto) [reply src/material/autocomplete/autocomplete.ts:365]: I see. I'm surprised that this didn't come up for `disableRipple` or `multiple` who follow the same pattern.

File: src/material/autocomplete/autocomplete.ts [hunk 5]
```diff
@@ -336,4 +343,37 @@ export class MatAutocomplete extends _MatAutocompleteBase {
...
+  }
+  private _hideSingleSelectionIndicator: boolean;
+
+  /** Syncs the parent state with the individual options. */
+  _syncParentProperties(): void {
+    if (this.options) {
+      Promise.resolve().then(() => {
```
Comment (crisbeto) [inline src/material/autocomplete/autocomplete.ts:372]: Why do we need to wrap this in a `Promise.resolve`? I wouldn't have expected it to trigger a "changed after checked" error.

File: src/material/autocomplete/autocomplete.ts [hunk 6]
```diff
@@ -336,4 +343,37 @@ export class MatAutocomplete extends _MatAutocompleteBase {
...
+
+  constructor(
+    _changeDetectorRef: ChangeDetectorRef,
+    _elementRef: ElementRef<HTMLElement>,
+    @Inject(MAT_AUTOCOMPLETE_DEFAULT_OPTIONS) defaults: MatAutocompleteDefaultOptions,
+    platform?: Platform,
+  ) {
+    super(_changeDetectorRef, _elementRef, defaults, platform);
+
+    this._hideSingleSelectionIndicator = defaults.hideSingleSelectionIndicator ?? false;
```
[the remark this answers is not in the document]
Comment (zarend) [reply src/material/autocomplete/autocomplete.ts:355]: Ok. This PR doesn't add a new usage of DI. Deprecating a constructor parameter would be a breaking change, which would go in a major release. What should we do in the meantime?

I can make the defaults protected, which is how some of our other components do it.

```
  constructor(
    _changeDetectorRef: ChangeDetectorRef,
    _elementRef: ElementRef<HTMLElement>,
    @Inject(MAT_AUTOCOMPLETE_DEFAULT_OPTIONS) protected _defaults: MatAutocompleteDefaultOptions,
    platform?: Platform,
  ) {
```

File: src/material/autocomplete/autocomplete.ts (hunk 5 above)
Comment (zarend) [reply src/material/autocomplete/autocomplete.ts]: Ohh yeah that's right. We don't need the `Promise.resolve` here. I removed it.

File: src/material/autocomplete/autocomplete.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (crisbeto) [reply src/material/autocomplete/autocomplete.ts:355]: Which parameter was deprecated? As far as I can tell, the constructor signature is the same as before.

[the remark this answers is not in the document]
Comment (zarend) [reply src/material/autocomplete/autocomplete.ts:355]: AFAIK, switching to using `inject` would require deprecating the `defaults` parameter.

As the PR is currently implemented, nothing is deprecated. This PR implements the constructor of derived class to take same arguments as base classes. That's so the derived class can have access to the `MAT_AUTOCOMPLETE_DEFAULT_OPTIONS`.

The above snippet is a suggested alternative to implementing the constructor of the derived class. It adds the `protected` keyword so that the derived class can access the `MAT_AUTOCOMPLETE_DEFAULT_OPTIONS`.

[the remark this answers is not in the document]
Comment (crisbeto) [reply src/material/autocomplete/autocomplete.ts:355]: I think it's fine to just make it protected in the base class.

File: src/material/autocomplete/autocomplete.ts (hunk 4 above)
Comment (zarend) [reply src/material/autocomplete/autocomplete.ts:365]: I can reproduce this problem with `disableRipple`. I think this has been happening all along, but there haven't been any issues filed for it because `disableRipple` and `multiple` are changed very rarely in practice.

I added a `disableRipple` option to the autocomplete demo and here is a video of the reproduction. It's kinda hard to see in the video, so here are the steps that happen.

(Ripple starts off enabled.
1. Click "Alabama"
2. (displays ripple animation)
3.  Click "Alabama" again
4. (displays ripple animation)
5. Click "TOGGLE DISABLE RIPPLE"
6. Click "Alabama"
7.  (displays ripple animation. Is not supposed to display ripple animation)
8. Click "Alabama" again
9.  (does not display ripple animation)


https://user-images.githubusercontent.com/7720245/210895519-67462c69-2abb-4a5a-af5f-c0e5216b0d74.mov

File: src/material/autocomplete/autocomplete.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (zarend) [reply src/material/autocomplete/autocomplete.ts]: Thanks, I changed it to make the defaults protected in the base class, which means that we don't need to touch the constructor of the derived class anymore 👍

**Comment by zarend** [comment]
Since this was last reviewed, changed out it puts the checkmark on the right of the mat option. Give the pseudo-checkbox margin auto rather than giving the list item content `flex-grow: 1;`. Increasing the size of the list item content causes unintended layout changes.

```
diff --git a/src/material/core/option/option.scss b/src/material/core/option/option.scss                                                                           [37/1987]
index 366eca739..714712b39 100644
--- a/src/material/core/option/option.scss
+++ b/src/material/core/option/option.scss
@@ -59,11 +59,11 @@
   }

   .mat-pseudo-checkbox-minimal {
-    margin-left: mdc-list-variables.$side-padding;
+    margin-left: auto;
     flex-shrink: 0;

     [dir='rtl'] & {
-      margin-right: mdc-list-variables.$side-padding;
+      margin-right: auto;
       margin-left: 0;
     }
   }
@@ -94,8 +94,6 @@
     font-family: inherit;
     text-decoration: inherit;
     text-transform: inherit;
-
-    flex-grow: 1;
   }

   @include cdk.high-contrast(active, off) {
   ```
